sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala

```diff
@@ -133,12 +133,36 @@ class AstBuilder extends DataTypeAstBuilder with SQLConfHelper with Logging {
 
   private def visitCompoundBodyImpl(
       ctx: CompoundBodyContext,
-      label: Option[String]): CompoundBody = {
+      label: Option[String],
+      allowPrefixDeclare: Boolean): CompoundBody = {
     val buff = ListBuffer[CompoundPlanStatement]()
     ctx.compoundStatements.forEach(compoundStatement => {
       buff += visit(compoundStatement).asInstanceOf[CompoundPlanStatement]
     })
 
+    val compoundStatements = buff.toList
+
+    if (allowPrefixDeclare) {
+      val declareAfterPrefix = compoundStatements.dropWhile(
+          statement => statement.isInstanceOf[SingleStatement] &&
+          statement.asInstanceOf[SingleStatement].parsedPlan.isInstanceOf[CreateVariable])
+        .filter(_.isInstanceOf[SingleStatement])
+        .exists(_.asInstanceOf[SingleStatement].parsedPlan.isInstanceOf[CreateVariable])
```

Comment:
once we add `varName` and `lineNumber` parameters to the error message, you'll probably need to use `find` instead of `exists` to find the first wrong variable declaration and construct error message based on it.